packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -191,6 +197,24 @@ export class HardhatNode extends EventEmitter {
           }
         }
       }
+
+      if (config.chains.has(forkNetworkId)) {
+        hardforkActivations = config.chains.get(forkNetworkId)!.hardforkHistory;
+      } else if (defaultHardhatNetworkParams.chains.has(forkNetworkId)) {
+        hardforkActivations =
+          defaultHardhatNetworkParams.chains.get(
+            forkNetworkId
+          )!.hardforkHistory;
+      } else {
+        throw new InternalError(
+          `No hardfork history configured for chain ID ${forkNetworkId}`
+        );
+      }
+      hardforkActivations.forEach((block, hardforkName) => {
```

Comment:
I'd also validate this when needed, to make this less disruptive. This is not necessary, but my reasoning is as follows:

- The node is initialized in the first request to the provider.
- Making this validation in the very first request will lead to an HFs related error in a seemingly unrelated request, like `eth_chainId`.
- Making it when we need to use a the unsupported hardfork will show the error in a situation where the user has more context.

Reply:
update: This is already validated in `_selectHardfork` so it can be removed

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1928/commits/6889614eecae24597dae138bc80312a5d9cd46e2